### PR TITLE
Make exception errors in PowerShell able to associate with the right history entry

### PR DESCRIPTION
**Pull request by daxian-dbw**
## PR Summary

Make errors in PowerShell able to associate with the right history entry, by adjusting the `HistoryId` property of the `InvocationInfo` of an `ErrorRecord`.

## PR Context

`ErrorRecord.InvocationInfo.HistoryId` should point to the history entry which contains the command-line that triggered the error. This is true for errors stemmed from `Write-Error`, `$PSCmdlet.WriteError`, `$PSCmdlet.ThrowTerminatingError`, and parameter binding failures.

However, it's not the case for errors stemmed from `throw` statement and other exceptions, such as `1/0` or method call exceptions. For those exceptions, the `HistoryId` turns out to be `-1`, which cannot be used to associate with the command from history that triggered the exception. This becomes a problem to the feedback provider feature we are working on, which needs to reliably tell if an error is resulted by the command that was just done execution.

This PR is to address this issue, to make errors stemmed from exceptions also points to the right history command that triggered the exception.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.

### CODE REVIEW

File: test/powershell/Language/Scripting/ErrorHistoryId.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+'@
+        ## 2nd in history
+        $null = RunCommand $ps 'bar'
+        $ps.HadErrors | Should -BeTrue
+        $err = RunCommand $ps '$Error[0]'
+        $err.FullyQualifiedErrorId | Should -BeExactly "PSCmdlet.ThrowTerminatingError,bar"
+        $err.InvocationInfo.HistoryId | Should -Be 2
+    }
+
+    It "Error from 'Throw' has the right history Id - 1" {
```
Comment (SteveL-MSFT) [inline test/powershell/Language/Scripting/ErrorHistoryId.Tests.ps1:79]: This test and the next one are very similar, you should turn it into `-TestCases`, something like:

```powershell
It "Error from 'Throw' has the right history Id: <commandOne>" -TestCases @(
  @{ commandOne = '1+1'; CommandTwo = "throw 'abc'" }
  @{ commandOne = "function bar { throw 'abc' }"; commandTwo = 'bar' }
){
        param ($commandOne, $commandTwo)
        try {
            ## 1st in history
            $null = RunCommand $ps $commandOne
            ## 2nd in history
            $null = RunCommand $ps $commandTwo
        } catch {
            ## ignore the exception
        }

        $ps.HadErrors | Should -BeTrue
        $err = RunCommand $ps '$Error[0]'
        $err.FullyQualifiedErrorId | Should -BeExactly 'abc'
        $err.InvocationInfo.HistoryId | Should -Be 2
    }
```

File: test/powershell/Language/Scripting/ErrorHistoryId.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            ## ignore the exception
+        }
+
+        $ps.HadErrors | Should -BeTrue
+        $err = RunCommand $ps '$Error[0]'
+        $err.FullyQualifiedErrorId | Should -BeExactly 'abc'
+        $err.InvocationInfo.HistoryId | Should -Be 2
+    }
+
+    It "Error from parameter binding has the right history Id" {
```
Comment (SteveL-MSFT) [inline test/powershell/Language/Scripting/ErrorHistoryId.Tests.ps1:111]: Same here, should be turned into test cases

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: test/powershell/Language/Scripting/ErrorHistoryId.Tests.ps1 (hunk 1 above)
Comment (daxian-dbw) [reply test/powershell/Language/Scripting/ErrorHistoryId.Tests.ps1]: Updated using `-TestCases`.

File: test/powershell/Language/Scripting/ErrorHistoryId.Tests.ps1 (hunk 2 above)
Comment (daxian-dbw) [reply test/powershell/Language/Scripting/ErrorHistoryId.Tests.ps1]: Updated using `-TestCases`

**Comment by daxian-dbw** [comment]
@SteveL-MSFT I have addressed your feedback. Please take another look, thanks!

**Review by PaulHigin — APPROVED** [review]
LGTM

File: src/System.Management.Automation/engine/lang/parserutils.cs [hunk 3]
```diff
@@ -1973,6 +1973,22 @@ internal static void UpdateExceptionErrorRecordPosition(Exception exception, ISc
                 }
             }
         }
+
+        internal static void UpdateExceptionErrorRecordHistoryId(RuntimeException exception, ExecutionContext context)
+        {
+            InvocationInfo invInfo = exception.ErrorRecord.InvocationInfo;
+            if (invInfo is null || invInfo.HistoryId != -1)
```
Comment (SeeminglyScience) [inline src/System.Management.Automation/engine/lang/parserutils.cs:1980]:
```suggestion
            if (invInfo is not { HistoryId: -1 })
```

Subjectively clearer new syntax, feel absolutely free to just mark as resolved if you disagree (it does compile the same FYI)

**Review by SeeminglyScience — APPROVED** [review]
LGTM! One optional suggestion

**Review by SteveL-MSFT — APPROVED** [review]

**Pull request merged**